### PR TITLE
Add support for named capturing groups

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
https://github.com/tc39/proposal-regexp-named-groups

I started working on this feature because I needed it for implementing a babel plugin. I then decided to use another package which already transformsnamed groups in regexps (regexp-tree), but I finished this PR anyway since it will probably be useful to someone else.

### CODE REVIEW

**Comment by jviereck** [comment]
Hi @nicolo-ribaudo,

thanks a lot for implementing this and proving the PR - especially as you went with using a different parser in the end :)

Doing a quick look over the code, is the named-group feature enabled by default? When adding unicode parsing before (see [1]), we've added a `features` flag to the `parse(...)` function. I think having a feature flag for this feature would be great as well. What do you think?

If you have any questions, let me know.

Thanks! \jv
 
[1]: https://github.com/jviereck/regjsparser/pull/78#issuecomment-220850082

**Comment by nicolo-ribaudo** [comment]
I enabled named groups by default because the only thing missing for the proposal to be merged to the main spec is the final approval, so it would be enabled by default[1] really soon ([relevant meeting notes](https://github.com/rwaldron/tc39-notes/blob/master/es8/2017-11/nov-28.md#9if-regexp-named-captures-status-update), https://github.com/tc39/ecma262/pull/1027). If you still prefer to enable this feature using a flag, let me know :wink: 

[1] What I say is only valid if the `features` option is only meant for experimental features.

**Comment by jviereck** [comment]
> f you still prefer to enable this feature using a flag, let me knowl

Yes, I prefer to have this feature enabled via a flag similar as in [1]. The reasoning is to avoid people using the library to undergo possible breaking/functionality change when adding this feature.

Thanks. If there is anything, let met know. \jv

[1]: https://github.com/jviereck/regjsparser/pull/78#issuecomment-220850082

**Comment by nicolo-ribaudo** [comment]
Done!

**Review by mathiasbynens — APPROVED** [review]

**Comment by jviereck** [comment]
hi folks,

First: Thanks @mathiasbynens for picking this one up again and asking me for review. I dropped the todo here. This is not the way it should be. Sorry @nicolo-ribaudo for delaying the review by this.

Overall: Thanks @nicolo-ribaudo for providing this PR. The code looks very clean and I like how you managed to follow the coding style of the library.

There are a few small nits, that I am going to comment code-inline next. 
@nicolo-ribaudo: Please let me know if you want to do the follow-up work yourself. Feel free to say no and I can take care of it ;)

If you have any questions, please let me know :),

All the best & thanks, 
\ julian

File: parser.js [hunk 1]
```diff
@@ -58,6 +58,7 @@
 //      DecimalEscape
 //      CharacterEscape
 //      CharacterClassEscape
+//      k GroupName
```
Comment (jviereck) [inline parser.js:61]: For understanding: the `k` is matching the character `k` here?

File: parser.js [hunk 2]
```diff
@@ -730,17 +794,77 @@
...
+      //      _
+      //      \ RegExpUnicodeEscapeSequence
+      //
+      // RegExpIdentifierContinue ::
+      //      UnicodeIDContinue
+      //      $
+      //      _
+      //      \ RegExpUnicodeEscapeSequence
+      //      <ZWNJ>
+      //      <ZWJ>
```
Comment (jviereck) [inline parser.js:829]: nit: shall we move the grammar for this up at the beginning of the file?

File: test/test-data-named-groups-unicode.json [hunk 3]
```diff
@@ -0,0 +1,60 @@
+{
```
Comment (jviereck) [inline test/test-data-named-groups-unicode.json:1]: Out of curiosity: Where did you/@nicolo-ribaudo get these tests from? Are these the official ECMA-262 tests for the named-groups feature?

File: test/index.js [hunk 4]
```diff
@@ -44,3 +44,9 @@ runTests(require('./test-data-unicode-properties.json'), 'u', {
   unicodePropertyEscape: true
 });
 runTests(require('./test-data-nonstandard.json'), '');
+runTests(require('./test-data-named-groups.json'), '', {
+  namedGroups: true
+});
+runTests(require('./test-data-named-groups-unicode.json'), 'u', {
+  namedGroups: true
+});
```
Comment (jviereck) [inline test/index.js:52]: How about adding tests for named groups & unicode-properties enabled jointly?

File: tools/generate-identifier-regex.js [hunk 5]
```diff
@@ -28,24 +28,24 @@ var generateES5Regex = function() { // ES 5.1
...
-  '// ECMAScript 5.1/Unicode v%s NonAsciiIdentifierPart:\n\n%s',
+  '// ECMAScript 5.1/Unicode v%s NonAsciiIdentifierStart:\n\n%s\n',
   version,
-  result.NonAsciiIdentifierPart
+  result.NonAsciiIdentifierStart
+);
+console.log(
+  '// ECMAScript 5.1/Unicode v%s NonAsciiIdentifierPartOnly:\n\n%s',
+  version,
+  result.NonAsciiIdentifierPartOnly
```
Comment (jviereck) [inline tools/generate-identifier-regex.js:50]: Ah, strange to `console.log` only one of the parts. Good catch :)

**Review by jviereck — COMMENTED** [review]
> There are a few small nits, that I am going to comment code-inline next.
>
> @nicolo-ribaudo: Please let me know if you want to do the follow-up work yourself. Feel free to say no and I can take care of it ;)

Done :)

**Comment by jviereck** [comment]
Do you mind if I wait with releasing a new version of the library once the jointly tests for the features are there?

File: parser.js (hunk 1 above)
Comment (nicolo-ribaudo) [reply parser.js:61]: Yes :+1:

File: test/test-data-named-groups-unicode.json (hunk 3 above)
Comment (nicolo-ribaudo) [reply test/test-data-named-groups-unicode.json:1]: No, I wrote them by myself.

File: tools/generate-identifier-regex.js (hunk 5 above)
Comment (nicolo-ribaudo) [reply tools/generate-identifier-regex.js:50]: Well, the other part wasn't used yet :slightly_smiling_face:

File: parser.js (hunk 2 above)
Comment (jviereck) [reply parser.js:829]: Looking over the code once more, other parts of the grammar are also defined locally. Therefore, leaving the grammar definition at the place introduced in this PR.
